File: dotnet/examples/Pactman.NonprofitCheckPlus.Examples/Cases/Ex23TransientRetries.cs
using System;
using System.Threading.Tasks;
using Pactman.NonprofitCheckPlus.Configuration;
using Pactman.NonprofitCheckPlus.Dev;
using Pactman.NonprofitCheckPlus.Exceptions;

namespace Pactman.NonprofitCheckPlus.Examples.Cases;

/// <summary>
/// EX-23 — Transient failures and retries.
/// </summary>
/// <remarks>
/// An endpoint that fails twice and then succeeds, and the statuses that are never retried
/// however the policy is configured.
/// </remarks>
public sealed class Ex23TransientRetries : IExample
{
    /// <inheritdoc />
    public string Id => "ex-23";

    /// <inheritdoc />
    public string Title => "Retrying a transient 503, and the statuses that are never retried.";

    /// <inheritdoc />
    public async Task RunAsync(string[] args)
    {
        using var context = ExampleContext.WithFixtures();

        Output.Heading("The default policy");
        Output.Field("MaxRetries", RetryOptions.Default.MaxRetries);
        Output.Field("InitialDelay", RetryOptions.Default.InitialDelay);
        Output.Field("MaxDelay", RetryOptions.Default.MaxDelay);
        Output.Field("BackoffFactor", RetryOptions.Default.BackoffFactor);
        Output.Field("Jitter", RetryOptions.Default.Jitter);
        Output.Field("RetryableStatuses", string.Join(", ", RetryOptions.Default.RetryableStatuses));
        Output.Field("RespectRetryAfter", RetryOptions.Default.RespectRetryAfter);

        Output.Heading("A 503 that clears");

        // The fixture answers 503 twice and then succeeds, which is exactly the shape a
        // retry policy exists for.
        var result = await context.Client.Nonprofits.CheckAsync(Fixtures.ControlEins.TransientFailure);

        Output.Field("status", result.Status);
        Output.Field("organization", result.Nonprofit?.OrganizationName);
        Output.Field("the caller saw", "one successful call — the failures were absorbed");

        Output.Heading("The same call with retrying off");

        using var noRetry = context.Sibling(retry: RetryOptions.None);

        try
        {
            await noRetry.Nonprofits.CheckAsync(Fixtures.ControlEins.TransientFailure);
            Output.Field("result", "succeeded — the fixture's failure budget had reset");
        }
        catch (PactmanServerException error)
        {
            Output.Field("type", error.GetType().Name);
            Output.Field("status", error.Status);
            Output.Field("attempts", error.Attempts);
        }

        Output.Heading("Statuses that are never retried");

        // 400, 401, 403 and 404 will not become a different answer, so the SDK refuses
        // to retry them however the policy is written.
        using var stubborn = context.Sibling(
            retry: new RetryOptions { MaxRetries = 5, RetryableStatuses = new[] { 404, 500, 503 } });

        try
        {
            await stubborn.Nonprofits.CheckAsync(Fixtures.Eins.NoRecord);
            Output.Field("404", "returned as an empty result rather than an error");
        }
        catch (PactmanNotFoundException error)
        {
            Output.Field("404 listed as retryable", "still not retried");
            Output.Field("attempts", error.Attempts);
        }

        Output.Heading("Tuning for one request");
        Output.Bullet("client.Retry with { MaxRetries = 5 } — start from the policy in force.");
        Output.Bullet("RetryOptions.None — no retrying for this request.");
        Output.Bullet("A CancellationToken bounds the whole call, retries included.");

        Output.Note(
            "Full jitter randomizes each delay across the whole range, so clients that\n"
            + "failed together do not retry in lockstep and re-create the outage.");
    }
}

File: ruby/examples/ex_23_transient_retries.rb
# frozen_string_literal: true

# EX-23 — Transient network or server failure.
#
# Retries are on by default: two of them, exponential backoff from 0.5 seconds
# with full jitter, capped at 8 seconds per delay. Eligible are 429, 500, 502,
# 503, 504 and connection failures that produced no response.
#
# Never retried, whatever `retryable_statuses` contains: 400, 401, 403, 404, and
# anything rejected by local validation. Retrying a rejected API key just burns
# the same key three times; retrying a 404 cannot make a record exist.
#
# Run:  PACTMAN_API_KEY=... bundle exec ruby examples/ex_23_transient_retries.rb

require_relative "lib/fixture_api"
require_relative "lib/print"

NCP = Pactman::NonprofitCheckPlus
include Print

def monotonic_ms
  (Process.clock_gettime(Process::CLOCK_MONOTONIC) * 1000).round
end

FixtureApi.with_fixture_api do |client, base_url|
  transient = FixtureApi::CONTROL_EINS[:transient_failure]

  # The fixture endpoint answers 503 twice, then succeeds — a textbook transient
  # failure. Delays are shortened here so the example runs quickly; the defaults
  # are 0.5 seconds initial and 8 seconds maximum.
  heading "A 503 that clears on retry"

  started_at = monotonic_ms
  result = client.nonprofits.check(transient, retry: { max_retries: 3, initial_delay: 0.04, max_delay: 0.4 })

  field "status", result.status
  field "organization", result.nonprofit&.organization_name
  field "elapsed (ms)", monotonic_ms - started_at
  bullet "Two 503s were absorbed; the caller saw one successful result."
  bullet "Backoff grows exponentially and is jittered, so parallel clients scatter."

  heading "The same failure with retries disabled"

  begin
    client.nonprofits.check(transient, retry: false)
    puts "  Unexpectedly succeeded."
  rescue NCP::ServerError => e
    field "class", e.class.name
    field "status", e.status
    field "attempts", e.attempts
    field "api_message", e.api_message
  end

  heading "Failures that are never retried"

  # 404 — a definite answer. Retrying cannot change it.
  begin
    client.nonprofits.check(FixtureApi::EINS[:no_record], retry: { max_retries: 5, retryable_statuses: [404, 500] })
  rescue NCP::NotFoundError => e
    field "404 attempts", "#{e.attempts} — not retried even though 404 was listed"
  end

  # 401 — retrying a rejected credential achieves nothing.
  bad_key_client = NCP::Client.new(api_key: "obviously-not-a-real-key", base_url: base_url, retry: { max_retries: 3 })

  begin
    bad_key_client.nonprofits.check(FixtureApi::EINS[:public_charity])
  rescue NCP::AuthenticationError => e
    field "401 attempts", "#{e.attempts} — authentication failures are terminal"
  end

  # Local validation — nothing was sent, so there is nothing to retry.
  begin
    client.nonprofits.check("not-an-ein", retry: { max_retries: 3 })
  rescue NCP::ValidationError => e
    field "validation", "origin=#{e.origin} — rejected before any request"
  end

  # A connection that never reaches a server: retried, then surfaced as a network
  # error carrying the attempt count and the underlying exception as its cause.
  heading "A connection failure"

  unreachable = NCP::Client.new(api_key: ENV.fetch("PACTMAN_API_KEY"), base_url: "http://127.0.0.1:1", timeout: 2,
                                retry: { max_retries: 2, initial_delay: 0.02, max_delay: 0.06 })

  begin
    unreachable.nonprofits.check(FixtureApi::EINS[:public_charity])
  rescue NCP::NetworkError => e
    field "class", e.class.name
    field "category", e.category.inspect
    field "attempts", e.attempts
    field "cause", e.cause.class.name
    field "message", e.message
  end
end

note "A retried failure that eventually succeeds is a success. A retried failure that\n" \
     "exhausts its budget is an outage — record it as \"not checked\", never as a pass."
